### ISSUE TITLE
CA1801 and CA1802 Incorrectly Flag C# Program Using Top-Level Statements

### ISSUE THREAD

**Issue by JasonBock** [dotnet/roslyn-analyzers#4184]
### Analyzer
CA1801 and CA1802

### Analyzer source
**NuGet Package**: [Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

**Version**: v3.3.0 (Latest)

### Describe the bug
I'm getting these issues with a type that is synthesized by the compiler and I have no control over. Top-level statements in C#9 creates a `<Program>$` class with a `<Main>$` method that isn't used in code.

### Steps To Reproduce
Create a .NET 5.0 console application (FYI I'm targeting .NET 5.0 RC1). Make sure the .csproj file looks like this:
```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <LangVersion>preview</LangVersion>
    <OutputType>Exe</OutputType>
    <TargetFramework>net5.0</TargetFramework>
    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.CodeAnalysis.FxCopAnalyzers" Version="3.3.0">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>
  </ItemGroup>
</Project>
```
Set the Program.cs file like this:
```
using System;

Console.WriteLine("hello");
```

### Expected behavior
The program prints "hello".

### Actual behavior
I get these warnings, which are errors because of the setting in the .csproj file:
```
Severity	Code	Description	Project	File	Line	Suppression State
Error	CA1801	Parameter args of method <Main>$ is never used. Remove the parameter or use it in the method body.	ConsoleApp6	C:\Users\jason\source\repos\ConsoleApp6\ConsoleApp6\CSC	1	Active
Error	CA1812	<Program>$ is an internal class that is apparently never instantiated. If so, remove the code from the assembly. If this class is intended to contain only static members, make it static (Shared in Visual Basic).	ConsoleApp6	C:\Users\jason\source\repos\ConsoleApp6\ConsoleApp6\Program.cs	3	Active
```
Note that I'm also seeing this issue in the Error List:
```
Severity	Code	Description	Project	File	Line	Suppression State
Error	AD0001	Analyzer 'Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines.IdentifiersShouldNotContainTypeNames' threw an exception of type 'System.IndexOutOfRangeException' with message 'Index was outside the bounds of the array.'.	ConsoleApp6	C:\Users\jason\source\repos\ConsoleApp6\ConsoleApp6\CSC	1	Active
```
I'd expect the analyzers to ignore this case as it's something the C# is doing and is a valid condition.

**Comment by mavasani** [comment]
Duplicate of https://github.com/dotnet/roslyn-analyzers/issues/4068 and https://github.com/dotnet/roslyn-analyzers/issues/4052

**Comment by kentyman23** [comment]
The title of this is wrong, as it should be mentioning 18**_1_**2 as shown in the description.
